Fix NoneType error when pulling non existent field

bolkedebruin opened the pull request:
If normalizing a jsonstruct a field can be absent
due to a schema change.

- [X] closes #30148
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

WillAyd commented:
Can you add test(s)? Typically the first thing we look for in a PR

bolkedebruin commented:
@WillAyd of course. Test has been added.

WillAyd commented:
Thanks for the test. I don't really agree with the change though - why wouldn't you want this to raise a `TypeError`?

bolkedebruin commented:
Imagine if I have 3 records:
```
[{'state': 'Florida', 'info': [1, 2,]}
[{'state': 'Texas', 'info': None}
[{'state': 'Texas', 'info': [2,3,]}
```

I cannot use json_normalize now with `record_path=['info']` as the field for the second record is None. That's perfectly valid for this field, but not accepteable to pandas. Its a real world case btw.

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -463,6 +463,11 @@ def test_nested_flattening_consistent(self):
+    def test_nonetype_record_path(self):
+        # see gh-30148
+        # should not raise TypeError
+        json_normalize([{'state': 'Texas', 'info': None}], record_path=['info'])
```
Can you assign the result of this expression to a variable named `result` then explicitly test what this should return via `tm.assert_frame_equal(result, expected)`? Should see this in other tests

bolkedebruin replied:
Sure

WillAyd commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -774,7 +774,7 @@ I/O
 - Bug in :class:`PythonParser` where str and bytes were being mixed when dealing with the decimal field (:issue:`29650`)
 - :meth:`read_gbq` now accepts ``progress_bar_type`` to display progress bar while the data downloads. (:issue:`29857`)
--
+- Bug in :func:`pandas.io.json.json_normalize` where a change of schema in json could lead to a NoneType error (:issue:`30148`)
```
```suggestion
- Bug in :func:`pandas.io.json.json_normalize` where a missing value in the location specified by `record_path` would raise a ``TypeError`` (:issue:`30148`)
```

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -463,6 +463,11 @@ def test_nested_flattening_consistent(self):
         # They should be the same.
         tm.assert_frame_equal(df1, df2)
 
+    def test_nonetype_record_path(self):
```
Can you parametrize this using the `nulls_fixture` fixture?

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -286,12 +286,16 @@ def _recursive_extract(data, path, seen_meta, level=0):
+                        else r
+                        for r in recs
+                    ]
+                    if recs
```
Can you move this outside of the generator expression for readability?

WillAyd reviewed (changes requested):
OK I think I see the point. Few things to improve this

bolkedebruin commented:
@WillAyd I have changed it to use the fixture, but I'm not entirely happy with it. The fixture sets it to float('NaN') which is different from 'None'. Now I am testing against an iterable and catching a `TypeError` which changes the gist of it.

What do you think?

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -286,12 +286,15 @@ def _recursive_extract(data, path, seen_meta, level=0):
+                    recs = [
+                        nested_to_record(r, sep=sep, max_level=max_level)
+                        if isinstance(r, dict)
+                        else r
```
dont' use large try/excepts, catch in nested_to_record

bolkedebruin replied:
@jreback `recs` can be non iterable. How to catch that in `nested_to_record`?

jreback replied:
huh? really? you can check if its is_scalar. but does this meet the guarantees?

bolkedebruin replied:
I had to use the `nulls_fixture` which makes it a scalar (see comment above). I rather use `None` which makes it cleaner and more appropriate imho.

jreback replied:
@bolkedebruin you are missing my point. so it seems that _pull_field can return a scalar and not just an iterable.

can you add a doc-string there (types as well if you can).
then just use is_scalar to detect this

bolkedebruin replied:
@jreback A gotcha. Yes I can do that. Still my comment above stands, I consider using `nulls_fixture` here a bit awkward: the `TypeError` original thrown was due a `None` value. Now I am also testing against a scalar? Is that the right thing? I’ll take you word for it, but I miss some understanding now.

I will update the code to test for `if recs and not is_scalar(xx)`.

jreback replied:
you reference ``nulls_fixture``, not sure what that has to do with anything. These are values which can be shown as null, they are scalars.

bolkedebruin replied:
For the test I was asked to use the nulls_fixture. This fixture sets the the value in the test to `float(‘NaN’)` which in this case is inherently (in my opinion) different from `None` and changes the nature of the test and the code to pass the test. In other words I do not understand the need for the `nulls_fixture` In this particular case, but maybe I am missing something. Using it does complicate the code and the changes you are requesting from me are related to that increased complexity.

jreback replied:
@bolkedebruin these are valid values in a dataframe. they must work. i don't think the code is more complex. using a giant try/except is not good practice generally as it tends to suppress things.

bolkedebruin replied:
Both points understood and not argued. I just wonder why this issue wasn’t discovered earlier. I’ll update the PR.

jreback replied:
> Both points understood and not argued. I just wonder why this issue wasn’t discovered earlier. I’ll update the PR.

no tests that hit the issue. ok thanks.

jreback commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -463,6 +463,15 @@ def test_nested_flattening_consistent(self):
         tm.assert_frame_equal(df1, df2)
 
+    def test_nonetype_record_path(self, nulls_fixture):
+        # see gh-30148
```
can you add an additional example that has some valid data in the result frame, it is not clear whether this should or should not do anything

bolkedebruin replied:
Will do

bolkedebruin commented:
@jreback @WillAyd ready for re-review

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -234,7 +235,10 @@ def _pull_field(js, spec):
         else:
-            result = result[spec]
+            try:
+                result = result[spec]
```
huh? again I really really don't like try/excepts they hide all kinds of errors. library code is NOT like user code where try/excepting is the norm. we *must* be very explicit.

bolkedebruin replied:
I gotcha entirely, but I was struggling how to accomplish this here. `is_integer` doesn't cut it here as `"1"` is valid for the list bit is not an integer according `is_integer`. If you have a suggestion that gets rid of this I'll gladly drop this.

jreback replied:
what case are you trying to get here? e.g. what is result and what is the spec?

bolkedebruin replied:
@jreback I think I got it. Thanks for pushing for the right fix btw.

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -286,12 +290,15 @@ def _recursive_extract(data, path, seen_meta, level=0):
-                    else r
-                    for r in recs
-                ]
+                if not is_scalar(recs):
```
it might be better for _pull_field to actually *always* return a list rather than a scalar, no? then you don't need this at all

bolkedebruin replied:
Good point. Will do.

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -234,7 +235,10 @@ def _pull_field(js, spec):
             for field in spec:
                 result = result[field]
         else:
```
adding type annotations here would go a *long* way towards clarifying what things are

bolkedebruin replied:
Will do.

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +228,17 @@ def json_normalize(
     """
 
-    def _pull_field(js, spec):
+    def _pull_field(js: Dict, spec: Union[List, str]) -> List:
```
```js: Dict[str, Any] ```

simonjayhawkins replied:
can type parameters be added to the generic List types?

jreback reviewed (changes requested):
looks good. ping when green.

bolkedebruin commented:
@jreback ping

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -228,13 +228,17 @@ def json_normalize(
             for field in spec:
-                result = result[field]
+                _res = _res[field]
+            result = cast(List, _res)
```
this is very odd to do a cast here. instead of this why don't you just do

```
result: List
```
at the top

jreback replied:
cc @simonjayhawkins

bolkedebruin replied:
@jreback the cast is here to satisfy mypi. It is unable to determine the right type otherwise (it will assume a dict) the cast is a no-op at runtime

simonjayhawkins replied:
> this is very odd to do a cast here. instead of this why don't you just do
> 
> ```
> result: List
> ```
> 
> at the top

I _think_ that it should be `result: Any` at the top. the isinstance check will allow mypy to narrow the types and ensure the correct return type.

bolkedebruin replied:
Given that `str` is acceptable as a return from `js[field]` I will change it to `result: Any` If the tests pass.

jreback commented on pandas/io/json/_normalize.py:
```diff
@@ -228,13 +228,17 @@ def json_normalize(
-            result = result[spec]
+            result = js[spec]
+
+        if not isinstance(result, Iterable):
```
this is even more strange

jreback replied:
if the tests don't fail then I am puzzled what exactly you are changing here

bolkedebruin replied:
This test fails (or are you talking about other tests?) when e.g. the nulls_fixtures is used. It will be a float then (float('NaN')). As a 'str' I also iterable (And valid according to the tests), I'm testing for this.

It was how I could make the existing tests pass and the newly added.

simonjayhawkins replied:
Should this be if not `isinstance(result, list)` or should the return type of _pull_field be typed as Iterable?

bolkedebruin replied:
I’ll change the return type to `Iterable` (`str` is valid and is not a list)

jreback commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -463,6 +463,19 @@ def test_nested_flattening_consistent(self):
+            ],
+            record_path=["info"],
+        )
+        df2 = DataFrame({"i": 2}, index=[0])
```
use result= and expected =
tm.assert_equal(result, expected)

bolkedebruin replied:
Done. This is different btw from the test above this test.

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,13 +228,17 @@ def json_normalize(
-        result = js
+    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> List:
         if isinstance(spec, list):
+            _res = js
```
What is the point of this assignment here? I think without this don't need all of the mypy discussion below

bolkedebruin replied:
@WillAyd Actually, the opposite was true. I set `result: List` which made it incompatible with `result = result[field]` (`result` was set to `js` above. `js` is not List and neither always iterable)

WillAyd replied:
What's the point of assigning this here and then overwriting it on line 235 within the loop?

bolkedebruin replied:
I just tested your solution. Tests do not pass. A `TypeError` is thrown.

bolkedebruin replied:
@WillAyd the loop drills down into the js. `_res` can be anything, `result` cannot be. `js[field]` can be anything hence I cannot assign `result` to `js[field]`

WillAyd replied:
Is it a TypeError on line 236? Why is the rhs of that referencing `_res` instead of `js`?

bolkedebruin replied:
It is a loop that goes over a nested `js`. Hence it reassigns it to do the drill down according to `spec`. 

```
    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> Iterable:
        result: Any = []
        if isinstance(spec, list):
            for field in spec:
>               result = _result[field]
E               TypeError: list indices must be integers or slices, not str
```

bolkedebruin replied:
I see that I made a mistake in my code to test it (I never did `result = js`)

bolkedebruin replied:
Let me quickly check that

bolkedebruin replied:
@WillAyd apologies, your sugestion does work. PR updated.

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +228,17 @@ def json_normalize(
         else:
             result = result[spec]
 
+        if not isinstance(result, Iterable):
```
Should this just be `pd.isnull(result)`? The consequence of this being an isinstance check would be that all scalar values would return nothing right? 

I know there was some conversation around this before but I think should disambiguate between missing values and scalar values; the former is fine whereas the latter should raise a `TypeError`

bolkedebruin replied:
You have a point. But should the `TypeError` be raised here then to make it more user friendly? Because otherwise the `TypeError` will be raised from 

```
                recs = [
                    nested_to_record(r, sep=sep, max_level=max_level)
                    if isinstance(r, dict)
                    else r
                    for r in recs
                ]
```

Which is not very indicative  of what is wrong and more importantly where It took me quite some time to find out in my data what went wrong and why as it can just happen in the middle of a 100mb file.

WillAyd replied:
Yea I think can raise a TypeError here if not missing or Dict/List

bolkedebruin replied:
so I will do a check with `pd.isnull` and will check with `is_scalar`.

WillAyd replied:
Yea I think that would be better

bolkedebruin replied:
*sigh* `str` is defined as a scalar. I'm not seeing away around to do `is_scalar(result) and not isinstance(result, Iterable)`. Or is there a better way?

WillAyd replied:
`if is_scalar(result) and pd.isnull(result)`?

bolkedebruin replied:
when `pd_isnull(result) == true` I need to `result = []`. `pd_isnull` doesn't deal with Iterables :-). I have updated the PR with a suggestion.

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +228,17 @@ def json_normalize(
-    def _pull_field(js, spec):
-        result = js
+    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> Iterable:
+        result: Any = js
```
Is this still necessary?

bolkedebruin replied:
What part? Stronger typing is always preferred I guess? That was suggested by @jreback . If I remove the `Any` I get

```
pandas/io/json/_normalize.py:240: error: Incompatible types in assignment (expression has type "List[<nothing>]", variable has type "Dict[str, Any]")
```

WillAyd replied:
Yea typing as `Any` isn't strong and for practical purposes would be the same as no typing.

`result: Union[List, Dict[str, Any]]` would be preferable for now; can refine later

bolkedebruin replied:
Wont work unfortunately. See https://github.com/python/mypy/issues/2220

```
pandas/io/json/_normalize.py:238: error: No overload variant of "__getitem__" of "list" matches argument type "str"
pandas/io/json/_normalize.py:238: note: Possible overload variants:
pandas/io/json/_normalize.py:238: note:     def __getitem__(self, int) -> Any
pandas/io/json/_normalize.py:238: note:     def __getitem__(self, slice) -> List[Any]
```

Adding `str` gives
```
error: Invalid index type "str" for "Union[str, List[Any], Dict[str, Any]]"; expected type "Union[int, slice]"
```

bolkedebruin replied:
If you don't mind I'll keep it at `Any` as that also benefits readability (imho)

WillAyd replied:
Yea the way this variable is reused makes things kind of tricky; not from anything you have here just a historical artifact

Instead of `: Any` can you just add a `# type: ignore` comment? The former we mostly use internally when an object really can be anything. That isn't true here, but just need some refactor to get it to work cleanly in a follow up and the ignore would be more indicative for that

bolkedebruin replied:
Sure

WillAyd reviewed (changes requested):
Glad the last thing worked; definitely makes things simpler. A few more details to hash out but going in the right direction

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -3,13 +3,14 @@
 from pandas._libs.writers import convert_json_to_lines
 
-from pandas import DataFrame
+from pandas import DataFrame, isnull
```
Would rather `import pandas as pd` as less prone to circular imports

bolkedebruin replied:
Okay. I'm assuming that you don't expect a refactor for this only for the function.

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +229,23 @@ def json_normalize(
+            if isnull(result):
+                result = []  # type: ignore
+            elif is_scalar(result):
+                raise TypeError(
```
Can you add a test case that this would hit?

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +229,23 @@ def json_normalize(
+        if not isinstance(result, Iterable):
+            if isnull(result):
+                result = []  # type: ignore
+            elif is_scalar(result):
```
On second read might be able to get rid of the `elif` and just keep to an `else`

bolkedebruin committed:
Fix TypeError when pulling field that is None

If normalizing a jsonstruct a field can be set to None
due to a schema change.

Co-Authored-By: William Ayd <william.ayd@icloud.com>

WillAyd commented on pandas/io/json/_normalize.py:
```diff
@@ -228,14 +229,23 @@ def json_normalize(
+                result = []  # type: ignore
+            else:
+                raise TypeError(
+                    "{} has non iterable value {} for path {}. "
```
If you do another iteration would be preferable to use f-strings here

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -463,6 +463,30 @@ def test_nested_flattening_consistent(self):
+        test_input = {"state": "Texas", "info": 1}
+        test_path = "info"
+        msg = (
+            "{} has non iterable value 1 for path {}. "
```
Same comment on f-strings

bolkedebruin replied:
done

WillAyd reviewed (approved):
lgtm. One comment that's not a blocker for me but others may disagree

Over to you @jreback

bolkedebruin commented:
Ping @jreback

bolkedebruin commented:
Friendly reminder @jreback

jreback commented:
can you merge master and will look again

bolkedebruin commented:
@jreback rebased. Or do you truly want me to merge master into my branch?

TomAugspurger commented on pandas/io/json/_normalize.py:
```diff
@@ -229,14 +230,23 @@ def _json_normalize(
-    def _pull_field(js, spec):
-        result = js
+    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> Iterable:
+        result = js  # type: ignore
```
Why do you need the `type: ignore` here?

bolkedebruin replied:
See: https://github.com/pandas-dev/pandas/pull/30145#discussion_r358483282 and the thread above

TomAugspurger reviewed (commented):
Merge master. There are some CI failures that should be fixed by doing that.

bolkedebruin committed:
Merge branch 'master' into fix_schema_change

bolkedebruin commented:
@jreback all green again.

jreback commented:
thanks @bolkedebruin very nice
